refactor: enforce single param on sending/editing methods

adrifcastr opened the pull request:
**Please describe the changes this PR makes and why it should be merged:**
This PR refactors the methods across the the lib's sending/editing methods to avoid inconsistent overloads by removing multiple param options and enforcing single param usage as laid out in #5771.

(and let's pretend that there isn't a typo in the branch name)
**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I pretend to know how to update typings and have done so, or typings don't need updating
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
<!--
Please move lines that apply to you out of the comment:
- This PR changes the library's interface (methods or parameters added)
- This PR **only** includes non-code changes, like changes to documentation, README, etc.
-->

ImRodry commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -52,25 +52,24 @@ class InteractionResponses {
    * Creates a reply to this interaction.
-   * @param {string|APIMessage|MessageAdditions} content The content for the reply
-   * @param {InteractionReplyOptions} [options] Additional options for the reply
+   * @param {InteractionReplyOptions} [options] The options for the reply
```
Wouldn't this be mandatory then?

ImRodry commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -131,12 +129,11 @@ class InteractionResponses {
    * Send a follow-up message to this interaction.
-   * @param {string|APIMessage|MessageAdditions} content The content for the reply
-   * @param {InteractionReplyOptions} [options] Additional options for the reply
+   * @param {InteractionReplyOptions} [options] The options for the reply
```
Same here

ImRodry commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -168,18 +165,17 @@ class InteractionResponses {
    * Updates the original message whose button was pressed
-   * @param {string|APIMessage|MessageAdditions} content The content for the reply
-   * @param {WebhookEditMessageOptions} [options] Additional options for the reply
+   * @param {WebhookEditMessageOptions} [options] The options for the update
```
And here

ImRodry reviewed (changes requested):
I don't see a reason for this to be done, but either way the docs are wrong

monbrey commented:
I like this change, it had been in the back of my mind for a while. I'd argue that we should do the same thing to `<TextChannel>.send` though, and enforce an object everywhere.

adrifcastr commented:
> I like this change, it had been in the back of my mind for a while. I'd argue that we should do the same thing to `<TextChannel>.send` though, and enforce an object everywhere.

Yes same, refer to internals, I got the go from Crawl to refactor the rest, hence why I changed the title and converted to draft

adrifcastr reviewed (commented):
-refactored again, comment dismissed-

DTrombett commented on src/structures/Message.js:
```diff
@@ -647,25 +646,22 @@ class Message extends Base {
+      APIMessage.transformOptions(options.content, options, {
+        reply: {
+          messageReference: this,
+          failIfNotExists: options?.failIfNotExists ?? options.content?.failIfNotExists ?? true,
```
I think the optional chaining operator here is not needed anymore because `options` is now required and it's not used in line 659 when accessing `options.content`. Also `options.content` is not an object but a string so `options.content?.failIfNotExists` makes no sense, I guess it was there before because the parameter `content` could be an instance of `APIMessage` too.

```suggestion
          failIfNotExists: options.failIfNotExists ??  true,
```

ckohen commented:
I *think* this would call for a refactor of `APIMessage.create` as well, since it would always be an options object, splitting it out no longer makes sense to me.

adrifcastr commented:
> I _think_ this would call for a refactor of `APIMessage.create` as well, since it would always be an options object, splitting it out no longer makes sense to me.

Actually probably not as I am currently refactoring to what the maintainers have internally decided upon, crawl laid it out in details in #5771

ckohen commented:
I did read the linked issue, it suggests only content should ever be provided standalone. meaning all the checks in `APIMessage.transformOptions` should never be hit, the whole purpose of those checks is to allow what is being removed here.
There will need to be a check whether options is a string though.

Something for the maintainers to address, should we still be able to pass an APIMessage to these methods still? I think it can be quite useful, but not sure if its wanted.

You can also just pass options to `APIMessage.create` already anyways, no need to `options.content, options`

adrifcastr commented:
Yeah I got a go on what to do, it's been discussed to pretty much do what the issue says and merge it into one param, and yeah I saw that the apimessage creation param is optional, will adress that :)

iCrawl commented on src/structures/Message.js:
```diff
@@ -541,18 +541,27 @@ class Message extends Base {
-   * @param {?(string|APIMessage)} [content] The new content for the message
-   * @param {MessageEditOptions|MessageEmbed|MessageAttachment|MessageAttachment[]} [options] The options to provide
+   * @param {string|APIMessage|MessageEditOptions|MessageEmbed
+   * |MessageAttachment|MessageAttachment[]} options The options to provide
```
Not sure if `MessageEmbed | MessageAttachment | MessageAttachment[]` should be here, wouldn't this imply that `message.edit(embed)` works?

ckohen replied:
Currently it does work, that's what I was trying to point out in https://github.com/discordjs/discord.js/pull/5758#issuecomment-855452913 So removing these from typings and docs can work, but leaves undocumented behavior (which people shouldn't use, but will)

iCrawl replied:
Yes, I was implying that this is undesired behaviour and should be fixed/removed (not only the docstring).

adrifcastr replied:
> Yes, I was implying that this is undesired behaviour and should be fixed/removed (not only the docstring).

Have done so within the latest commits

iCrawl commented on typings/index.d.ts:
```diff
@@ -1247,11 +1244,7 @@ declare module 'discord.js' {
-    public edit(
-      content: string | null,
-      options: MessageEditOptions | MessageEmbed | MessageAttachment | MessageAttachment[],
+      content: string | MessageEditOptions | MessageEmbed | APIMessage | MessageAttachment | MessageAttachment[],
```
Same as above.

iCrawl commented on typings/index.d.ts:
```diff
@@ -1362,20 +1343,16 @@ declare module 'discord.js' {
     public deleteReply(): Promise<void>;
     public editReply(
-      content: string | APIMessage | WebhookEditMessageOptions | MessageEmbed | MessageEmbed[],
+      options: string | APIMessage | WebhookEditMessageOptions | MessageEmbed,
```
Somewhat same as above?

iCrawl commented on typings/index.d.ts:
```diff
@@ -1362,20 +1343,16 @@ declare module 'discord.js' {
+    public reply(options: string | APIMessage | InteractionReplyOptions | MessageAdditions): Promise<void>;
     public update(
-      content: string | APIMessage | WebhookEditMessageOptions | MessageEmbed | MessageEmbed[],
+      content: string | APIMessage | WebhookEditMessageOptions | MessageEmbed,
```
Same as above.

iCrawl commented on typings/index.d.ts:
```diff
@@ -2134,23 +2111,15 @@ declare module 'discord.js' {
       message: MessageResolvable,
-      content: string | null | APIMessage | MessageEmbed | MessageEmbed[],
-      options?: WebhookEditMessageOptions,
+      options: string | APIMessage | WebhookEditMessageOptions | MessageEmbed | MessageEmbed[],
```
Same as above.

ckohen commented on src/structures/Message.js:
```diff
@@ -541,18 +541,27 @@ class Message extends Base {
+
+    if (options instanceof APIMessage) {
+      data = options.resolveData();
+    } else if (typeof options === 'string') {
```
This check isn't necessary with the current APIMessage, it's done in APIMessage.transformOptions (in reverse) currently. I think I may have confused you when I said you should check for strings. What I was trying to say is that transformOptions should perform this check and only ever be passed options and not content, options

adrifcastr replied:
Thank you :), adressed

ckohen commented on src/structures/Message.js:
```diff
@@ -647,26 +656,36 @@ class Message extends Base {
+
+    if (options instanceof APIMessage) {
+      data = options;
+    } else if (typeof options === 'string') {
```
ditto

ckohen commented on src/structures/Webhook.js:
```diff
@@ -146,13 +146,15 @@ class Webhook {
-      apiMessage = content.resolveData();
+    if (options instanceof APIMessage) {
+      apiMessage = options.resolveData();
+    } else if (typeof options === 'string') {
```
ditto

ckohen commented on src/structures/Webhook.js:
```diff
@@ -242,15 +244,19 @@ class Webhook {
+    let apiMessage;
+
+    if (options instanceof APIMessage) apiMessage = options.resolveData();
+    else if (typeof options === 'string') apiMessage = APIMessage.create(this, options).resolveData();
```
ditto

ckohen commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -64,13 +63,18 @@ class InteractionResponses {
+
+    let apiMessage;
+    if (options instanceof APIMessage) apiMessage = options;
+    else if (typeof options === 'string') apiMessage = APIMessage.create(this, options);
```
ditto

ckohen commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -100,17 +104,20 @@ class InteractionResponses {
+    if (typeof options === 'string' || options instanceof APIMessage) {
+      return this.webhook.editMessage('@original', options);
+    } else {
+      return this.webhook.editMessage('@original', null, options);
```
This method only takes options now, same as InteractionWebhook.editMessage does, it should remain a passthrough, simply removing the content parameter.

ckohen commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -129,12 +136,12 @@ class InteractionResponses {
-    return this.webhook.send(content, options);
+  followUp(options) {
+    if (typeof options === 'string' || options instanceof APIMessage) return this.webhook.send(options);
+    else return this.webhook.send(null, options);
```
Same as above

ckohen commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -158,18 +165,22 @@ class InteractionResponses {
+
+    let apiMessage;
+    if (options instanceof APIMessage) apiMessage = options;
+    else if (typeof options === 'string') apiMessage = APIMessage.create(this, options);
```
ditto

ckohen commented on src/structures/interfaces/TextBasedChannel.js:
```diff
@@ -156,20 +155,22 @@ class TextBasedChannel {
 
     if (this instanceof User || this instanceof GuildMember) {
-      return this.createDM().then(dm => dm.send(content, options));
+      return this.createDM().then(dm => dm.send(null, options));
```
```suggestion
      return this.createDM().then(dm => dm.send(options));
```

ckohen commented on src/structures/interfaces/TextBasedChannel.js:
```diff
@@ -156,20 +155,22 @@ class TextBasedChannel {
-      apiMessage = content.resolveData();
+    if (options instanceof APIMessage) {
+      apiMessage = options.resolveData();
+    } else if (typeof options === 'string') {
```
ditto

ckohen commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -64,13 +63,18 @@ class InteractionResponses {
+    let apiMessage;
+    if (options instanceof APIMessage) apiMessage = options;
+    else if (typeof options === 'string') apiMessage = APIMessage.create(this, options);
+    else apiMessage = APIMessage.create(this, null, options);
```
are you intentionally passing options as extras here?

adrifcastr replied:
Not intentionally no, thank you, will fix

ckohen commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -100,17 +104,20 @@ class InteractionResponses {
+    if (typeof options === 'string' || options instanceof APIMessage) {
+      return this.webhook.editMessage('@original', options);
+    } else {
+      return this.webhook.editMessage('@original', null, options);
```
This would result in an API Error

adrifcastr replied:
Did this before refactoring APIMessage so it slipped right past me

ckohen commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -158,18 +164,22 @@ class InteractionResponses {
+    let apiMessage;
+    if (options instanceof APIMessage) apiMessage = options;
+    else if (typeof options === 'string') apiMessage = APIMessage.create(this, options);
+    else apiMessage = APIMessage.create(this, null, options);
```
same question

vaporox commented on src/structures/Message.js:
```diff
@@ -541,18 +541,22 @@ class Message extends Base {
-  edit(content, options) {
-    options = content instanceof APIMessage ? content : APIMessage.create(this, content, options);
-    return this.channel.messages.edit(this.id, options);
+  edit(options) {
+    let apiMessage;
+
+    if (options instanceof APIMessage) apiMessage = options.resolveData();
+    else apiMessage = APIMessage.create(this, options).resolveData();
+
+    return this.channel.messages.edit(this.id, apiMessage);
   }
```
```suggestion
  edit(options) {
    return this.channel.messages.edit(this.id, options);
  }
```
The additional logic here was added in #5612, not sure if it is still needed after this PR but I believe not

vaporox commented on src/structures/Webhook.js:
```diff
@@ -132,7 +131,8 @@ class Webhook {
    * // Send an embed with a local image inside
-   * webhook.send('This is an embed', {
+   * webhook.send({
+   * content: 'This is an embed',
```
```suggestion
   *   content: 'This is an embed',
```

vaporox commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -64,13 +63,18 @@ class InteractionResponses {
-    const apiMessage = content instanceof APIMessage ? content : APIMessage.create(this, content, options);
+
+    let apiMessage;
+    if (options instanceof APIMessage) apiMessage = options;
+    else if (typeof options === 'string') apiMessage = APIMessage.create(this, options);
+    else apiMessage = APIMessage.create(this, options);
```
```suggestion
    if (options instanceof APIMessage) apiMessage = options;
    else apiMessage = APIMessage.create(this, options);
```
The `else if` and `else` seem to do exactly the same, not sure if I'm missing something

adrifcastr replied:
Yes, thanks, must've also slipped through when redoing it after the initial PR

vaporox commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -100,17 +104,20 @@ class InteractionResponses {
    */
-  editReply(content, options) {
-    return this.webhook.editMessage('@original', content, options);
+  editReply(options) {
+    if (typeof options === 'string' || options instanceof APIMessage) {
+      return this.webhook.editMessage('@original', options);
+    } else {
+      return this.webhook.editMessage('@original', options);
+    }
   }
```
```suggestion
  editReply(options) {
    return this.webhook.editMessage('@original', options);
  }
```
Same as above

vaporox commented on src/structures/interfaces/InteractionResponses.js:
```diff
@@ -158,18 +164,22 @@ class InteractionResponses {
-    const apiMessage = content instanceof APIMessage ? content : APIMessage.create(this, content, options);
+
+    let apiMessage;
+    if (options instanceof APIMessage) apiMessage = options;
+    else if (typeof options === 'string') apiMessage = APIMessage.create(this, options);
+    else apiMessage = APIMessage.create(this, options);
```
```suggestion
    if (options instanceof APIMessage) apiMessage = options;
    else apiMessage = APIMessage.create(this, options);
```
Same as above

vaporox commented on typings/index.ts:
```diff
@@ -38,9 +38,9 @@ client.on('message', ({ channel }) => {
+  assertIsMessage(channel.send({ files: [attachment] }));
   assertIsMessage(channel.send(embed));
-  assertIsMessage(channel.send([attachment, embed]));
+  assertIsMessage(channel.send({ embed: embed, files: [attachment] }));
```
```suggestion
  assertIsMessage(channel.send({ embed, files: [attachment] }));
```

vladfrangu commented on src/structures/Message.js:
```diff
@@ -647,26 +646,28 @@ class Message extends Base {
    * Send an inline reply to this message.
-   * @param {string|APIMessage} [content=''] The content for the message
-   * @param {ReplyMessageOptions|MessageAdditions} [options] The additional options to provide
+   * @param {string|APIMessage|ReplyMessageOptions} options The options to provide
```
So edit can take a MessageEmbed, but reply can't? :/

vladfrangu commented on typings/index.d.ts:
```diff
@@ -1142,16 +1120,14 @@ declare module 'discord.js' {
       options: InteractionReplyOptions | APIMessage,
     ): Promise<Message | RawMessage | (Message | RawMessage)[]>;
     public send(
       content: string | null,
```
This can never happen

adrifcastr replied:
Should be resolved within the latest commits

vladfrangu commented on typings/index.d.ts:
```diff
@@ -1142,16 +1120,14 @@ declare module 'discord.js' {
+      options: InteractionReplyOptions & { split?: false },
     ): Promise<Message | RawMessage>;
     public send(
       content: string | null,
```
This can never happen

adrifcastr replied:
Should also be resolved within the latest commits

SpaceEEC commented on typings/index.d.ts:
```diff
@@ -2487,12 +2414,8 @@ declare module 'discord.js' {
-    send(content: string | null, options: MessageOptions & { split: true | SplitOptions }): Promise<Message[]>;
-    send(content: string | null, options: MessageOptions): Promise<Message | Message[]>;
+    send(options: string | (MessageOptions & { split?: false })): Promise<Message>;
+    send(options: string | (MessageOptions & { split: true | SplitOptions })): Promise<Message[]>;
```
Same here too:
```suggestion
    send(options: APIMessage (MessageOptions & { split: true | SplitOptions })): Promise<Message[]>;
```

SpaceEEC commented on typings/index.d.ts:
```diff
@@ -2533,31 +2456,13 @@ declare module 'discord.js' {
-    send(
-      content: string | null,
-      options: WebhookMessageOptions & { split: true | SplitOptions },
+      options: string | APIMessage | (WebhookMessageOptions & { split: true | SplitOptions }),
```
And here too:
```suggestion
      options: APIMessage | (WebhookMessageOptions & { split: true | SplitOptions }),
```

SpaceEEC commented on typings/index.d.ts:
```diff
@@ -1261,20 +1220,8 @@ declare module 'discord.js' {
-      options: ReplyMessageOptions & { split: true | SplitOptions },
-    ): Promise<Message[]>;
-    public reply(content: string | null, options: ReplyMessageOptions): Promise<Message | Message[]>;
+    public reply(options: string | (ReplyMessageOptions & { split?: false })): Promise<Message>;
```
APIMessage may be used here too:
```suggestion
    public reply(options: string | APIMessage | (ReplyMessageOptions & { split?: false })): Promise<Message>;
```

SpaceEEC commented on typings/index.d.ts:
```diff
@@ -2487,12 +2414,8 @@ declare module 'discord.js' {
-    send(content: string | null, options: (MessageOptions & { split?: false }) | MessageAdditions): Promise<Message>;
-    send(content: string | null, options: MessageOptions & { split: true | SplitOptions }): Promise<Message[]>;
-    send(content: string | null, options: MessageOptions): Promise<Message | Message[]>;
+    send(options: string | (MessageOptions & { split?: false })): Promise<Message>;
```
```suggestion
    send(options: string | APIMessage | (MessageOptions & { split?: false })): Promise<Message>;
```

SpaceEEC reviewed (dismissed):
Some things I found:

adrifcastr commented on typings/index.d.ts:
```diff
@@ -151,25 +151,10 @@ declare module 'discord.js' {
     public static create(
```
Thank you, removed in the recent commits, the overload didn't seem to serve any purpose, after my refactor of APIMessage anyway

adrifcastr commented on src/structures/Webhook.js:
```diff
@@ -242,15 +242,18 @@ class Webhook {
+  async editMessage(message, options) {
+    let apiMessage;
+
+    if (options instanceof APIMessage) apiMessage = options.resolveData();
```
Done in the latest commits, thank you

SpaceEEC commented on typings/index.d.ts:
```diff
@@ -2134,23 +2064,13 @@ declare module 'discord.js' {
     public send(
-      content: string | null,
-      options: WebhookMessageOptions & { split: true | SplitOptions },
+      options: string | APIMessage | (WebhookMessageOptions & { split: true | SplitOptions }),
```
Passing a string can't be spit.
```suggestion
      options: APIMessage | (WebhookMessageOptions & { split: true | SplitOptions }),
```

adrifcastr replied:
Thank you, honestly split is a very confusing part in all of this, especially the typings

SpaceEEC commented on typings/index.d.ts:
```diff
@@ -1143,24 +1114,11 @@ declare module 'discord.js' {
     public send(
-      content: string | null,
-      options: InteractionReplyOptions & { split: true | SplitOptions },
+      options: string | APIMessage | (InteractionReplyOptions & { split: true | SplitOptions }),
```
Same here
```suggestion
      options: APIMessage | (InteractionReplyOptions & { split: true | SplitOptions }),
```

SpaceEEC reviewed (changes requested):
Seems like I missed two:

AustrianDev commented:
Some examples were not changed :
https://github.com/discordjs/discord.js/blob/7b2e12b102984abf61132e1057558ef7f04e6d83/src/structures/interfaces/TextBasedChannel.js#L146-L158
I saw another but it got patched while I was writing this comment lol

AustrianDev commented:
same here :
https://github.com/discordjs/discord.js/blob/7b2e12b102984abf61132e1057558ef7f04e6d83/src/structures/interfaces/InteractionResponses.js#L167-L170

kyranet commented:
Are you able to open a PR fixing those docs, @AustrianDev?

If so, it'd be nice if you could work on those 👍🏼

AustrianDev commented:
Ok, I will
